ENIGMAsystem/SHELL/Graphics_Systems/Direct3D9/Direct3D9Headers.h

```diff
@@ -15,10 +15,197 @@
+
+#include <map>
 #include <string>
+#include <cmath>
+
+#include <d3d9.h>
+#include <windows.h>
+#include <windowsx.h>
+
 using namespace std;
-#endif
+
+using std::map;
+
+namespace enigma {
+
+extern HWND hWnd;
+extern bool Direct3D9Managed;
+
+namespace dx9 {
+
+class ContextManager {
+private:
+
+protected:
+map< D3DRENDERSTATETYPE, DWORD > cacheRenderStates;                  /// cached RenderStates
+map< DWORD, LPDIRECT3DTEXTURE9 > cacheTextureStates;                 /// cached Texture States
+map< DWORD, map< D3DSAMPLERSTATETYPE, DWORD > > cacheSamplerStates;  /// cached Sampler States
+map< DWORD, D3DLIGHT9 > cacheLightStates;                            /// cached Light States
+map< DWORD, BOOL > cacheLightEnable;                                 /// cached Light States
+
+public:
+LPDIRECT3DDEVICE9 device;    // the pointer to the device class
+
+ContextManager() {
+
+}
+
+~ContextManager() {
+
+}
+
+// Reapply the render states and other stuff to the device.
+void RestoreState() {
+  // Cached Texture Stage States
+  map< DWORD, LPDIRECT3DTEXTURE9 >::iterator tit = cacheTextureStates.begin();
+  while (tit != cacheTextureStates.end()) {
+    device->SetTexture(tit->first, tit->second);
+    tit++;
+  }
+
+  // Cached Render States
+  map< D3DRENDERSTATETYPE, DWORD >::iterator it = cacheRenderStates.begin();
+  while (it != cacheRenderStates.end()) {
+    device->SetRenderState(it->first, it->second);
+    it++;
+  }
+
+  // Cached Sampler States
+  map< DWORD, map< D3DSAMPLERSTATETYPE, DWORD > >::iterator sit = cacheSamplerStates.begin();
+  while (sit != cacheSamplerStates.end()) {
+    map< D3DSAMPLERSTATETYPE, DWORD >::iterator secit = sit->second.begin();
+    while (secit != sit->second.end()) {
+      device->SetSamplerState(sit->first, secit->first, secit->second);
+      secit++;
+    }
+    sit++;
+  }
+
+  // Cached Lights
+  map< DWORD, BOOL >::iterator lite = cacheLightEnable.begin();
+  while (lite != cacheLightEnable.end()) {
+    device->LightEnable(lite->first, lite->second);
+    lite++;
+  }
+  return;
+  map< DWORD, D3DLIGHT9 >::iterator lit = cacheLightStates.begin();
+  while (lit != cacheLightStates.end()) {
+    device->SetLight(lit->first, &lit->second);
+    lit++;
+  }
+}
+
+void LightEnable(DWORD Index, BOOL bEnable) {
+  // Update the light state cache
+  // If the return value is 'true', the command must be forwarded to the D3D Runtime.
+  map< DWORD, BOOL >::iterator it = cacheLightEnable.find(Index);
+  if (cacheLightEnable.end() == it) {
+    cacheLightEnable.insert(map< DWORD, BOOL >::value_type(Index, bEnable));
+    device->LightEnable(Index, bEnable);
+    return;
+  }
+  if (it->second == bEnable)
+    return;
+  it->second = bEnable;
+  device->LightEnable(Index, bEnable);
+}
+
+void Release() {
+  device->Release();
+}
+
+void SetMaterial(const D3DMATERIAL9 *pMaterial) {
+  device->SetMaterial(pMaterial);
+}
+
+void SetLight(DWORD Index, const D3DLIGHT9 *pLight) {
```

Comment:
Did you copy this stuff from somewhere? And more importantly, is it public domain?

Reply:
No I did not copy it and I am the author.